s3/src/test/scala/akka/stream/alpakka/s3/auth/SignerSpec4.scala

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.s3.auth
+
+import java.io.File
+
+import akka.actor.ActorSystem
+import akka.stream.{ ActorMaterializer, ActorMaterializerSettings, Materializer }
+import akka.testkit.TestKit
+import org.scalatest.{ FlatSpecLike, Matchers }
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.time.{ Millis, Seconds, Span }
+import java.nio.file.{ Files, Path }
+import java.time.{ LocalDate, LocalDateTime, ZoneOffset, ZonedDateTime }
+
+import akka.http.scaladsl.model.ContentType.WithFixedCharset
+import akka.http.scaladsl.model.HttpHeader.ParsingResult
+import akka.http.scaladsl.model.{ Uri, _ }
+
+import scala.concurrent.Future
+import scala.io.Source
+import scala.util.{ Failure, Success, Try }
+import scala.language.implicitConversions
+
+/**
+ * Validates implementation of AWS Signature version 4.
+ */
+class SignerSpec4(_system: ActorSystem) extends TestKit(_system) with FlatSpecLike with ScalaFutures with Matchers {
+  def this() = this(ActorSystem("SignerSpec4"))
+
+  implicit val defaultPatience =
+    PatienceConfig(timeout = Span(2, Seconds), interval = Span(5, Millis))
+
+  implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system).withDebugLogging(true))
+
+  implicit def rawHttpRequest(source: String): HttpRequestFromSource = new HttpRequestFromSource(source)
+
+  import materializer.executionContext
+
+  val credentials = AWSCredentials("AKIDEXAMPLE", "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY")
+  val scope = CredentialScope(LocalDate.of(2015, 8, 30), "us-east-1", "service")
+  val signingKey = SigningKey(credentials, scope)
+
+  private val reqs = Probes.loadExpectedResults(RequestProbe)
+  private val creqs = Probes.loadExpectedResults(CanonicalRequestProbe)
+  private val stss = Probes.loadExpectedResults(StringToSignProbe)
+  private val authzs = Probes.loadExpectedResults(AuthorizationHeaderProbe)
+  private val sreqs = Probes.loadExpectedResults(SignedRequestProbe)
+
+  private val requests = reqs.map(_.map(_._2.fromSource))
+
+  it should "produce proper canonical requests" in {
+    val probes = for {
+      req <- requests
+      crq <- creqs
+    } yield req.zip(crq)
+
+    probes.toOption shouldBe defined
+
+    probes.get.foreach {
+      case (fileName, expectedResult) =>
+        testCanonicalRequest(fileName, expectedResult)
+    }
```

Comment:
All 4 test cases are different only with testMethod and expected result - can this be a single function with results collection and test function as arguments?